File: smart-farm-operations-platform-main/backend/src/main/java/com/smartfarm/features/activity/domain/Activity.java
package com.smartfarm.features.activity.domain;

import com.smartfarm.features.farm.domain.Farm;
import com.smartfarm.features.field.domain.Field;
import com.smartfarm.features.crop.domain.Crop;
import com.smartfarm.features.auth.domain.User;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import java.time.OffsetDateTime;
import java.util.UUID;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.SQLDelete;
import org.hibernate.annotations.SQLRestriction;
import org.hibernate.annotations.UpdateTimestamp;

@Entity
@Table(name = "activities")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@SQLDelete(sql = "UPDATE activities SET deleted = true, deleted_at = CURRENT_TIMESTAMP WHERE id = ?")
@SQLRestriction("deleted = false")
public class Activity {
    @Id
    @GeneratedValue
    private UUID id;

    @Column(nullable = false, length = 255)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(name = "activity_type", nullable = false, length = 50)
    private ActivityType activityType;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 50)
    @Builder.Default
    private ActivityStatus status = ActivityStatus.PLANNED;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 50)
    @Builder.Default
    private ActivityPriority priority = ActivityPriority.MEDIUM;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "farm_id", nullable = false)
    private Farm farm;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "field_id", nullable = false)
    private Field field;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "crop_id")
    private Crop crop;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "performed_by")
    private User performedBy;

    @Column(name = "scheduled_date", nullable = false)
    private OffsetDateTime scheduledDate;

    @Column(name = "completed_date")
    private OffsetDateTime completedDate;

    @Column(name = "estimated_duration")
    private Integer estimatedDuration;

    @Column(name = "actual_duration")
    private Integer actualDuration;

    @Column(columnDefinition = "TEXT")
    private String notes;

    @Column(columnDefinition = "TEXT")
    private String attachments;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by")
    private User createdBy;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "updated_by")
    private User updatedBy;

    @Column(name = "season")
    private String season;

    @Column(name = "start_date")
    private OffsetDateTime startDate;

    @Column(name = "end_date")
    private OffsetDateTime endDate;

    @Column(name = "estimated_cost", precision = 12, scale = 2)
    private java.math.BigDecimal estimatedCost;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "supervisor_id")
    private User supervisor;

    @Column(name = "required_equipment", columnDefinition = "TEXT")
    private String requiredEquipment;

    @Column(name = "required_inventory", columnDefinition = "TEXT")
    private String requiredInventory;

    @Builder.Default
    private boolean deleted = false;

    @Column(name = "deleted_at")
    private OffsetDateTime deletedAt;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private OffsetDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private OffsetDateTime updatedAt;
}

File: smart-farm-operations-platform-main/frontend/src/types/activity.ts
export type ActivityType =
  | 'IRRIGATION'
  | 'FERTILIZER'
  | 'PESTICIDE'
  | 'HARVEST'
  | 'PLANTING'
  | 'SOIL_TEST'
  | 'MAINTENANCE'
  | 'INSPECTION'
  | 'OTHER';

export type ActivityStatus = 'PLANNED' | 'IN_PROGRESS' | 'COMPLETED' | 'CANCELLED';

export type ActivityPriority = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';

export interface FarmActivity {
  id: string;
  title: string;
  description?: string;
  activityType: ActivityType;
  status: ActivityStatus;
  priority: ActivityPriority;
  farmId: string;
  farmCode: string;
  farmName: string;
  fieldId: string;
  fieldCode: string;
  fieldName: string;
  cropId?: string;
  cropName?: string;
  performedBy?: string;
  performedByName?: string;
  performedByPhone?: string;
  scheduledDate: string;
  completedDate?: string;
  estimatedDuration?: number;
  actualDuration?: number;
  notes?: string;
  attachments?: string;
  createdBy?: string;
  createdByName?: string;
  updatedBy?: string;
  updatedByName?: string;
  createdAt: string;
  updatedAt: string;
  season?: string;
  startDate?: string;
  endDate?: string;
  estimatedCost?: number;
  supervisorId?: string;
  supervisorName?: string;
  requiredEquipment?: string;
  requiredInventory?: string;
  progress?: number;
}
